## A "Hello World" Example of Machine Learning - Revisit

Loading the Iris dataset from scikit-learn. 

The first column represents Sepal length, the second column represents Sepal width,  the third column represents the petal length, and the fourth column the petal width of the flower samples. The classes (type of species) are already converted to integer labels where 0=Iris-Setosa, 1=Iris-Versicolor, 2=Iris-Virginica.

Here, we are using only two features: the third and fourth columns. 

In [987]:
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
#iris.data

In [988]:
X = iris.data[:, [0, 3]]

y = iris.target

In [989]:
print('Class labels:', np.unique(y))

Class labels: [0 1 2]


Scikit-learn algorithms support multi-class classification via the One-Versus-Rest(OvR) method. 

Splitting data into 70% training and 30% test data:

In [990]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

In [991]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]


### Standardizing the features:

In [992]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [993]:
from sklearn.linear_model import Perceptron

ppn = Perceptron(max_iter=100, eta0=0.1, random_state=42)
ppn.fit(X_train_std, y_train)

Perceptron(eta0=0.1, max_iter=100, random_state=42)

### Test the model with the hold-out test set

In [994]:
y_pred = ppn.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 10


In [995]:
from sklearn.metrics import accuracy_score

print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.7777777777777778


In [996]:
X_new = [[1.1, 0.2],[0.4, 1.9], [1.4, 0.2]]
y_new = ppn.predict(X_new)
y_new

array([2, 2, 2])

### Evaluate the model using cross validation

In [997]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.88888889, 0.76923077, 0.53846154, 0.80769231])

2-features: sccuracy score: array([0.88888889, 0.48148148, 0.7037037 , 0.95833333])

### Exercise 1: Use all four features to train the model and use cross validaton to check if the results better? Briefly explain why. 

In [998]:
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
X = iris.data[:, [0,1,2,3]]
y = iris.target
print('Class labels:', np.unique(y))


Class labels: [0 1 2]


In [ ]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

In [999]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

In [1000]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [1002]:
from sklearn.linear_model import Perceptron

ppn = Perceptron(max_iter=100, eta0=0.1, random_state=42)
ppn.fit(X_train_std, y_train)

Perceptron(eta0=0.1, max_iter=100, random_state=42)

In [1003]:
y_pred = ppn.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 3


In [1004]:
from sklearn.metrics import accuracy_score

print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.9333333333333333


In [1005]:
X_new = [[5.1,3.5,1.4,0.2],[4.9,3.0,1.4,0.2], [4.7,3.2,1.3,0.2]]
y_new = ppn.predict(X_new)
y_new

array([1, 1, 1])

In [1006]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.92592593, 0.88461538, 0.84615385, 0.88461538])

### Accuracy for 2 features = 0.7777777777777778
### cross_value_score for 2 features =array([0.88888889, 0.76923077, 0.53846154, 0.80769231])

### Accuracy for 4 features =  0.9333333333333333
### cross_value_score for 4 features = array([0.92592593, 0.88888889, 0.81481481, 0.91666667])

#### therefore accuracy is high with 4-features than 2-features and also cross_value_score also increases for 4-features, 
output is better with four features.

### Exercise 2: Try with the scikit-learn stochastic gradient descent model instead of perceptron. Use all four features. Evaluate with cross-validation how does the model perform in terms of accuracy using both two features and four features. 

With 2 features

In [1007]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier(random_state=42)

In [1008]:
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
#iris.data

In [1009]:
X = iris.data[:, [1,3]]

In [1010]:
print('Class labels:', np.unique(y))

Class labels: [0 1 2]


In [1011]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

In [1012]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]


In [1013]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [1014]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train_std, y_train)

SGDClassifier(random_state=42)

In [1015]:
y_pred = sgd.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 1


In [1016]:
from sklearn.metrics import accuracy_score

print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.9777777777777777


In [1017]:
X_new = [[5.1,3.5,1.4,0.2],[4.9,3.0,1.4,0.2], [4.7,3.2,1.3,0.2]]
y_new = ppn.predict(X_new)
y_new

array([1, 1, 1])

In [1018]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.88888889, 0.84615385, 0.80769231, 0.80769231])

With 4 features


In [1020]:
from sklearn import datasets
import numpy as np

iris = datasets.load_iris()
#iris.data

In [1021]:
X = iris.data[:, [0,1,2,3]]

In [1022]:
print('Class labels:', np.unique(y))

Class labels: [0 1 2]


In [1023]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=1, stratify=y)

In [1024]:
print('Labels counts in y:', np.bincount(y))
print('Labels counts in y_train:', np.bincount(y_train))
print('Labels counts in y_test:', np.bincount(y_test))

Labels counts in y: [50 50 50]
Labels counts in y_train: [35 35 35]
Labels counts in y_test: [15 15 15]


In [1025]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler() #center the distribution around zero (mean), with a standard deviation of 1.
sc.fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [1026]:
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
sgd = SGDClassifier(random_state=42)
sgd.fit(X_train_std, y_train)

SGDClassifier(random_state=42)

In [1027]:
y_pred = sgd.predict(X_test_std)
print('Misclassified samples: ' + str((y_test != y_pred).sum()))

Misclassified samples: 5


In [1028]:
from sklearn.metrics import accuracy_score

print('Accuracy: ' + str(accuracy_score(y_test, y_pred)))

Accuracy: 0.8888888888888888


In [1029]:
X_new = [[5.1,3.5,1.4,0.2],[4.9,3.0,1.4,0.2], [4.7,3.2,1.3,0.2]]
y_new = ppn.predict(X_new)
y_new

array([1, 1, 1])

In [1030]:
from sklearn.model_selection import cross_val_score
cross_val_score(ppn, X_train_std, y_train, cv=4, scoring="accuracy")

array([0.92592593, 0.88461538, 0.84615385, 0.88461538])

In [ ]:
Accuracy for 2 features =:0.9777777777777777
array for two features([0.88888889, 0.84615385, 0.80769231, 0.80769231])

Accuracy for four features: 0.8888888888888888
array for four features([0.92592593, 0.88461538, 0.84615385, 0.88461538])

